src/apps/external_executor/external_executor.cpp

```diff
@@ -45,6 +52,35 @@ namespace externalexecutor
         ccf::no_auth_required)
         .set_auto_schema<void, GetExecutorCode::Out>()
         .install();
+
+      // create an endpoint to register the executor
+      auto register_executor = [this](auto& ctx, ccf::NewExecutor&& payload)
+        -> ccf::grpc::GrpcAdapterResponse<ccf::RegistrationResult> {
+        // generate and store executor node id locally
+        crypto::Pem executor_public_key(payload.cert());
```

Comment:
I think we should use the public key from the `ctx` object (I'm assuming we're doing mutual TLS here?). See existing example in `node_frontend.h` around line `440` and the use of `rpc_ctx->get_session_context()->caller_cert`.

Reply:
This is intentionally retrieved from the body. For a couple of reasons as discussed with @eddyashton 

1. We avoid the need to obtain a private key from the `Executor`
2. If we go the route of offloading registration to a side car, we don't end up passing a private key this way (although there is still a chance that the side car could go rogue on its own without a private key as well)

Reply:
To clarify: we're _not_ doing mutual TLS auth on the register endpoint - it's possible and expected that the registration is done over an anonymous TLS channel.

Another benefit of doing it this way that I hadn't quite twigged yesterday, is that the auth policy for registration is extremely awkward the other way round. You need to come with _a_ TLS cert, but not one we know anything about yet? In fact, if it is known, is that an auth error? Is a mismatch between the TLS cert and the attestation an auth error? I think it's simpler to reason about this cert alongside the attestation purely from the (transport-agnostic!) payload.

Reply:
We still need to check for ownership of the private key corresponding to the public key, otherwise an attacker could submit a copy the attestation of the valid node (assumed public), along with the valid node's public key and successfully register that way. We could guarantee ownership of the public key with a signature using the corresponding private key and submit that signature in the payload but this doesn't guarantee freshness like TLS does.

Reply:
Why do we need to check ownership at this stage? That allows an attacker to register a valid executor, but not to substitute their own key in any way - the result is still that the executor is registered, with the executor's public key. When they use that for TLS to actually interact (ie - make any of the calls to the `KV` service), they prove ownership of the corresponding private key at that point.

Reply:
OK, that makes sense but e.g. if the external executor gets recorded in the ledger (I don't think it is in the current plan?), then it may at least confusing to an auditor?

Reply:
Yes in the current plan, executor details are local to the node and won't be in the ledger.

Reply:
@jumaffre even if we did, that would be indistinguishable from an executor that had registered successfully, but then never handled any requests, perhaps because it crashed in between. This is true both in terms of audit and impact.